File: typescript/src/polling.ts
import type { PollingOptions } from './types';

export class PollingTimeoutError extends Error {
  constructor(message = 'Polling timed out') {
    super(message);
    this.name = 'PollingTimeoutError';
  }
}

export class PollingMaxRetriesError extends Error {
  constructor(retries: number) {
    super(`Polling exceeded maximum retries (${retries})`);
    this.name = 'PollingMaxRetriesError';
  }
}

/**
 * Polls a function until a condition is met, using exponential backoff.
 *
 * @param fn - Async function to call on each poll
 * @param isComplete - Predicate that returns true when polling should stop
 * @param options - Polling configuration
 * @returns The final result from `fn` when `isComplete` returns true
 */
export async function pollUntilComplete<T>(
  fn: () => Promise<T>,
  isComplete: (result: T) => boolean,
  options?: PollingOptions,
): Promise<T> {
  const {
    initialDelay = 1000,
    maxDelay = 30000,
    maxRetries = 60,
    backoffMultiplier = 2,
    timeout,
    signal,
  } = options ?? {};

  const startTime = Date.now();
  let delay = initialDelay;
  let retries = 0;

  while (true) {
    if (signal?.aborted) {
      throw new DOMException('Polling was aborted', 'AbortError');
    }

    const result = await fn();

    if (isComplete(result)) {
      return result;
    }

    retries++;

    if (retries >= maxRetries) {
      throw new PollingMaxRetriesError(maxRetries);
    }

    if (timeout !== undefined && Date.now() - startTime >= timeout) {
      throw new PollingTimeoutError();
    }

    await sleep(delay, signal);
    delay = Math.min(delay * backoffMultiplier, maxDelay);
  }
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    if (signal?.aborted) {
      reject(new DOMException('Polling was aborted', 'AbortError'));
      return;
    }

    const timer = setTimeout(resolve, ms);

    signal?.addEventListener(
      'abort',
      () => {
        clearTimeout(timer);
        reject(new DOMException('Polling was aborted', 'AbortError'));
      },
      { once: true },
    );
  });
}

File: python/ecf_dgii/polling.py
"""Polling utilities with exponential backoff for ECF processing."""

from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass, field
from typing import Any

from .exceptions import PollingMaxRetriesError, PollingTimeoutError


@dataclass
class PollingOptions:
    """Configuration for polling ECF processing status.

    Attributes:
        initial_delay: Seconds to wait before the first poll.
        max_delay: Maximum seconds between polls.
        max_retries: Maximum number of poll attempts (0 = unlimited).
        backoff_multiplier: Multiplier applied to delay each iteration.
        timeout: Total timeout in seconds (0 = unlimited).
    """

    initial_delay: float = 2.0
    max_delay: float = 30.0
    max_retries: int = 0
    backoff_multiplier: float = 1.5
    timeout: float = 300.0


TERMINAL_PROGRESS = {"Completed", "Failed", "Rejected"}


async def poll_until_complete(
    poll_fn: Any,
    *,
    options: PollingOptions | None = None,
) -> Any:
    """Call *poll_fn* repeatedly until the result indicates completion.

    ``poll_fn`` must be an async callable that returns an object with a
    ``progress`` attribute (or dict key). Polling stops when ``progress``
    is in a terminal state.
    """
    opts = options or PollingOptions()
    delay = opts.initial_delay
    retries = 0
    start = time.monotonic()

    while True:
        result = await poll_fn()

        progress = getattr(result, "progress", None)
        if progress is None and isinstance(result, dict):
            progress = result.get("progress")

        progress_value = progress.value if hasattr(progress, "value") else str(progress)

        if progress_value in TERMINAL_PROGRESS:
            return result

        if opts.timeout and (time.monotonic() - start) >= opts.timeout:
            raise PollingTimeoutError(
                f"Polling timed out after {opts.timeout}s (last progress: {progress_value})"
            )

        retries += 1
        if opts.max_retries and retries >= opts.max_retries:
            raise PollingMaxRetriesError(
                f"Polling exceeded {opts.max_retries} retries (last progress: {progress_value})"
            )

        await asyncio.sleep(delay)
        delay = min(delay * opts.backoff_multiplier, opts.max_delay)
